File: apps/api/src/modules/proxy/provider-resolver.ts
import type { Env } from "@raven/config";
import type { Database } from "@raven/db";
import { providerConfigs } from "@raven/db";
import { and, eq } from "drizzle-orm";
import type { Redis } from "ioredis";
import { cachedQuery } from "@/lib/cache-utils";
import { decrypt } from "@/lib/crypto";
import {
  ForbiddenError,
  NotFoundError,
  UnauthorizedError,
  ValidationError
} from "@/lib/errors";
import { type RoutingStrategy, resolveWithStrategy } from "./router";

export interface ProviderResolution {
  decryptedApiKey: string;
  providerConfigId: string;
  providerName: string;
  upstreamPath: string;
}

export const parseProviderFromPath = (
  reqPath: string
): { providerName: string; configId: string | null; upstreamPath: string } => {
  const pathSegments = reqPath.replace(/^\/v1\/proxy\/?/, "").split("/");
  const providerSegment = pathSegments[0] ?? "";
  const tildeIdx = providerSegment.indexOf("~");
  return {
    configId: tildeIdx === -1 ? null : providerSegment.slice(tildeIdx + 1),
    providerName:
      tildeIdx === -1 ? providerSegment : providerSegment.slice(0, tildeIdx),
    upstreamPath: `/${pathSegments.slice(1).join("/")}`
  };
};

export const resolveProvider = async (
  db: Database,
  env: Env,
  organizationId: string,
  reqPath: string,
  redis?: Redis,
  strategy?: RoutingStrategy
): Promise<ProviderResolution> => {
  const { providerName, configId, upstreamPath } =
    parseProviderFromPath(reqPath);

  if (!providerName) {
    throw new ValidationError("Provider not specified in path");
  }

  let providerConfig: typeof providerConfigs.$inferSelect | null = null;

  if (configId) {
    const queryFn = async () => {
      const [result] = await db
        .select()
        .from(providerConfigs)
        .where(
          and(
            eq(providerConfigs.id, configId),
            eq(providerConfigs.organizationId, organizationId),
            eq(providerConfigs.provider, providerName)
          )
        )
        .limit(1);
      return result ?? null;
    };
    providerConfig = redis
      ? await cachedQuery(redis, `pc:id:${configId}`, 60, queryFn)
      : await queryFn();
  } else if (redis) {
    const resolvedId = await resolveWithStrategy(
      db,
      redis,
      organizationId,
      providerName,
      strategy
    );
    providerConfig = await cachedQuery(
      redis,
      `pc:id:${resolvedId}`,
      60,
      async () => {
        const [result] = await db
          .select()
          .from(providerConfigs)
          .where(eq(providerConfigs.id, resolvedId))
          .limit(1);
        return result ?? null;
      }
    );
  } else {
    const allConfigs = await db
      .select()
      .from(providerConfigs)
      .where(
        and(
          eq(providerConfigs.organizationId, organizationId),
          eq(providerConfigs.provider, providerName),
          eq(providerConfigs.isEnabled, true)
        )
      )
      .limit(10);

    providerConfig =
      allConfigs[Math.floor(Math.random() * allConfigs.length)] ?? null;
  }

  if (!providerConfig) {
    throw new NotFoundError(
      configId
        ? `No provider config found for '${providerName}' with ID '${configId}'`
        : `No provider config found for '${providerName}'`
    );
  }

  if (!providerConfig.isEnabled) {
    throw new ForbiddenError(`Provider '${providerName}' is disabled`);
  }

  let decryptedApiKey: string;
  try {
    decryptedApiKey = decrypt(providerConfig.apiKey, env.ENCRYPTION_SECRET);
  } catch {
    throw new UnauthorizedError("Failed to decrypt provider credentials");
  }

  return {
    decryptedApiKey,
    providerConfigId: providerConfig.id,
    providerName,
    upstreamPath
  };
};

File: apps/api-py/src/proxy/provider_resolver.py
from redis.asyncio import Redis
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from src.config import settings
from src.lib.crypto import decrypt_with_rotation
from src.lib.errors import AppError
from src.models.provider import ProviderConfig


def parse_provider_from_path(path: str) -> str:
    parts = path.strip("/").split("/")
    if parts:
        return parts[0]
    raise AppError("Invalid provider path", 400, "INVALID_PATH")


async def resolve_provider(
    session: AsyncSession, provider_path: str, redis: Redis
) -> dict:
    provider_name = parse_provider_from_path(provider_path)

    result = await session.execute(
        select(ProviderConfig).where(
            ProviderConfig.provider == provider_name,
            ProviderConfig.is_enabled.is_(True),
        )
    )
    config = result.scalar_one_or_none()

    if not config:
        raise AppError(
            f"No active provider configuration found for '{provider_name}'",
            404,
            "PROVIDER_NOT_FOUND",
        )

    decrypted_key = decrypt_with_rotation(
        config.api_key,
        settings.ENCRYPTION_SECRET,
        settings.ENCRYPTION_SECRET_PREVIOUS,
    )

    return {
        "api_key": decrypted_key,
        "config_id": config.id,
        "config_name": config.name,
        "provider_name": provider_name,
        "upstream_path": provider_path,
    }
